components/scream/src/share/grid/remap/horizontal_remap_utility.hpp

```diff
@@ -0,0 +1,564 @@
+    Kokkos::deep_copy(dofs_gids_h,m_dofs_gids);
+    for (int ii=0; ii<dofs_gids_h.extent(0); ii++) {
+      printf("%10d: %10d\n",ii,dofs_gids_h(ii));
+    } 
+    printf("\n=============================================\n");
+  }
+/*---------------------------------------------*/
+  void check() {
+    std::vector<bool> found(m_dofs_gids.size(),false);
+    auto dofs_gids_h = Kokkos::create_mirror_view(m_dofs_gids);
+    Kokkos::deep_copy(dofs_gids_h,m_dofs_gids);
+    for (int ii=0; ii<map_segments.size(); ii++) {
+      auto& seg = map_segments[ii];
+      auto seg_check = seg.check();
+      EKAT_REQUIRE_MSG(seg_check,"Error in GSMap " + map_name + " - problem with a remapping segment for dof = " + std::to_string(seg.m_dof) + ".");
+      for (int jj=0; jj<dofs_gids_h.size(); jj++) {
+        if (dofs_gids_h(jj) == seg.m_dof) {
+          found[jj] = true;
+          break;
+        }
+      }
+    }
+    bool pass = std::find(found.begin(),found.end(),false) == found.end();
+    if (!pass) {
+      printf("Error: GSMap.check() for map %s - Target column(s) are missing a remap segment:\n",map_name.c_str());
+      for (int ii=0;ii<m_dofs_gids.size();ii++) {
+        if (!found[ii]) {
+          printf("No segment found for DOF = %d\n",dofs_gids_h(ii));
+        }
+      }
+    }
+    EKAT_REQUIRE(pass);
+  };
+/*---------------------------------------------*/
+  // Useful if needing to grab source data from somewhere, like a data file.
+  void set_unique_source_dofs() {
+    // Make sure the unique_dofs vector is clear before appending it
+    unique_dofs.clear();
+    // Check all segments and add unique dofs.
+    for (int iseg=0; iseg<get_num_of_segs(); iseg++) {
+      const auto& seg = map_segments[iseg];
+      const auto& src_dofs = seg.source_dofs;
+      const auto& src_dofs_h = Kokkos::create_mirror_view(src_dofs);
+      Kokkos::deep_copy(src_dofs_h,src_dofs);
+      for (int ii=0; ii<seg.m_length; ii++) {
+        auto idx = std::find(unique_dofs.begin(), unique_dofs.end(), src_dofs_h(ii));
+        if (idx == unique_dofs.end()) {
+          unique_dofs.push_back(src_dofs_h(ii));
+        }
+      }
+    }
+    std::sort(unique_dofs.begin(), unique_dofs.end());
+    for (int iseg=0; iseg<get_num_of_segs(); iseg++) {
+      const auto& seg = map_segments[iseg];
+      const auto& src_dofs = seg.source_dofs;
+      const auto& src_dofs_h = Kokkos::create_mirror_view(src_dofs);
+      Kokkos::deep_copy(src_dofs_h,src_dofs);
+      auto& src_idx  = seg.source_idx;
+      auto  src_idx_h  = Kokkos::create_mirror_view(src_idx);
+      for (int ii=0; ii<seg.m_length; ii++) {
+        auto idx = std::find(unique_dofs.begin(), unique_dofs.end(), src_dofs_h(ii));
+        int idx_ii = idx - unique_dofs.begin();
+        src_idx_h(ii) = idx - unique_dofs.begin();
+      }
+      Kokkos::deep_copy(src_idx,src_idx_h);
+    }
+  }
+/*---------------------------------------------*/
+  RemapSegment get_segment_impl(const gid_type dof) {
```

Comment:
These `_impl` functions should probably be `private`, to keep from cluttering the interface.